### PR TITLE
8189441: Define algorithm names for keys derived from KeyAgreement

### PR DESCRIPTION
**Pull request by wangweij**
Allow `Generic` as an algorithm in the `KeyAgreement::generateSecret(alg)` method.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8347059](https://bugs.openjdk.org/browse/JDK-8347059) to be approved

### Issues
 * [JDK-8189441](https://bugs.openjdk.org/browse/JDK-8189441): Define algorithm names for keys derived from KeyAgreement (**Bug** - P4)
 * [JDK-8347059](https://bugs.openjdk.org/browse/JDK-8347059): Define algorithm names for keys derived from KeyAgreement (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22650/head:pull/22650` \
`$ git checkout pull/22650`

Update a local copy of the PR: \
`$ git checkout pull/22650` \
`$ git pull https://git.openjdk.org/jdk.git pull/22650/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22650`

View PR using the GUI difftool: \
`$ git pr show -t 22650`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22650.diff">https://git.openjdk.org/jdk/pull/22650.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22650#issuecomment-2528833736)
</details>

### CODE REVIEW

**Comment by wangweij** [comment]
The comparison of algorithm names was inconsistent, sometimes `equals` and sometimes `equalsIgnoreCase`. This is even more suspicious in the `P11KeyAgreement` where both were used. My understanding is that `equalsIgnoreCase` should always be used because key algorithm has always been case-insensitive. Also, since the "TlsPremasterSecret" algorithm is only used internally, it is always provided with the correct case.

File: src/java.base/share/classes/com/sun/crypto/provider/DHKeyAgreement.java [hunk 1]
```diff
@@ -419,10 +419,14 @@ protected SecretKey engineGenerateSecret(String algorithm)
                 throw new InvalidKeyException("Key material is too short");
             }
             return skey;
-        } else if (algorithm.equals("TlsPremasterSecret")) {
+        } else if (KeyUtil.isSupportedKeyAgreementOutputAlgorithm(algorithm)) {
             // remove leading zero bytes per RFC 5246 Section 8.1.2
```
Comment (seanjmullan) [inline src/java.base/share/classes/com/sun/crypto/provider/DHKeyAgreement.java:423]: Maybe move this comment after line 424 inside the `if` block.

File: test/jdk/java/security/KeyAgreement/Generic.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8189441
+ * @library /test/lib /test/jdk/sun/security/pkcs11
+ * @summary make sure Generic is accepted by all KeyAgreement
```
Comment (seanjmullan) [inline test/jdk/java/security/KeyAgreement/Generic.java:28]: Suggest: s/KeyAgreement/KeyAgreement implementations/

File: src/java.base/share/classes/javax/crypto/KeyAgreement.java [hunk 3]
```diff
@@ -659,9 +659,16 @@ public final int generateSecret(byte[] sharedSecret, int offset)
      * {@code generateSecret} to change the private information used in
      * subsequent operations.
      *
-     * @param algorithm the requested secret-key algorithm
-     *
-     * @return the shared secret key
+     * @param algorithm the requested secret-key algorithm. This is
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/KeyAgreement.java:662]: Did you consider instead referencing the new section you are adding for standard key algorithm names (https://bugs.openjdk.org/browse/JDK-8346736)? It could be reworded as:

"See the Key Algorithms section in the Java Security Standard Algorithm Names Specification for information about standard secret key algorithm names. Use "Generic" ..."

Comment (wangweij) [reply src/java.base/share/classes/javax/crypto/KeyAgreement.java:662]: Good idea.

Comment (wangweij) [reply src/java.base/share/classes/javax/crypto/KeyAgreement.java]: Do you suggest me adding the same to `encapsulate` and `decapsulate` of `KEM` in this PR?

Comment (seanjmullan) [reply src/java.base/share/classes/javax/crypto/KeyAgreement.java]: Yes.

**Comment by seanjmullan** [comment]
Can you also replace the link in `Key.getAlgorithm` with the new section you will be adding?

**Comment by wangweij** [comment]
> Can you also replace the link in `Key.getAlgorithm` with the new section you will be adding?

OK, and maybe more. I just found `SecretKeySpec` and `EncodedKeySpec`.

**Comment by wangweij** [comment]
I've modified too many files and I think they should belong to https://bugs.openjdk.org/browse/JDK-8346736. We should only touch KEM and KeyAgreement here.

**Comment by seanjmullan** [comment]
> I've modified too many files and I think they should belong to https://bugs.openjdk.org/browse/JDK-8346736. We should only touch KEM and KeyAgreement here.

Makes sense.

**Comment by wangweij** [comment]
I've reverted all changes not related to `KeyAgreement` and put them in https://github.com/openjdk/jdk/pull/22929, including the `KEM` one, since it involves no behavior change and is pure clarification. I've also removed the "KEM" word from the bug title.

File: src/java.base/share/classes/javax/crypto/KeyAgreement.java [hunk 4]
```diff
@@ -659,9 +659,19 @@ public final int generateSecret(byte[] sharedSecret, int offset)
...
      * subsequent operations.
      *
-     * @param algorithm the requested secret-key algorithm
+     * @param algorithm the requested secret-key algorithm. This is different
+     *      from the {@code KeyAgreement} algorithm provided to the
+     *      {@code getInstance} method. See the SecretKey Algorithms section in the
+     *      <a href="{@docRoot}/../specs/security/standard-names.html#secretkey-algorithms">
+     *      Java Security Standard Algorithm Names Specification</a>
+     *      for information about standard secret key algorithm names.
+     *      Use "Generic" if the output will be used as the input keying
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/KeyAgreement.java:668]: s/Use/Specify/

File: src/java.base/share/classes/javax/crypto/KeyAgreement.java [hunk 5]
```diff
@@ -659,9 +659,19 @@ public final int generateSecret(byte[] sharedSecret, int offset)
...
+     *      material of a key derivation function (KDF).
      *
-     * @return the shared secret key
+     * @return the shared secret key. The length of the key material
+     *      may be adjusted to be compatible with the specified algorithm,
+     *      regardless of whether the key is extractable. If {@code algorithm}
+     *      is specified as "Generic", the full shared secret will be returned.
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/KeyAgreement.java:674]: I think this last sentence is too strict, as some existing provider implementations may not support "Generic". I would suggest rewording as "If {@code algorithm} is specified as "Generic" and is supported by the implementation, the full shared secret is returned."

File: src/java.base/share/classes/sun/security/ec/XDHKeyAgreement.java [hunk 6]
```diff
@@ -207,9 +209,9 @@ protected SecretKey engineGenerateSecret(String algorithm)
             throw new NoSuchAlgorithmException("Algorithm must not be null");
         }
 
-        if (!(algorithm.equals("TlsPremasterSecret"))) {
+        if (!KeyUtil.isSupportedKeyAgreementOutputAlgorithm(algorithm)) {
             throw new NoSuchAlgorithmException(
-                    "Only supported for algorithm TlsPremasterSecret");
+                    "Unsupported secret key algorithm " + algorithm);
```
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/ec/XDHKeyAgreement.java:214]: Might be clearer to add a `:` after algorithm, also more consistent with DHKeyAgreement.

File: src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyAgreement.java [hunk 7]
```diff
@@ -268,20 +268,18 @@ protected SecretKey engineGenerateSecret(String algorithm)
             throw new NoSuchAlgorithmException("Algorithm must not be null");
         }
 
-        if (algorithm.equals("TlsPremasterSecret")) {
+        if (KeyUtil.isSupportedKeyAgreementOutputAlgorithm(algorithm)) {
             // For now, only perform native derivation for TlsPremasterSecret
-            // as that is required for FIPS compliance.
+            // as that is required for FIPS compliance and Generic for KDF.
```
Comment (seanjmullan) [inline src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyAgreement.java:273]: Suggest breaking into two sentences: "For now, only perform native derivation for TlsPremasterSecret and Generic algorithms. TlsPremasterSecret is required for FIPS compliance and Generic is required for input to KDF."

File: src/java.base/share/classes/javax/crypto/KeyAgreement.java [hunk 8]
```diff
@@ -659,9 +659,20 @@ public final int generateSecret(byte[] sharedSecret, int offset)
...
+     *      Java Security Standard Algorithm Names Specification</a>
+     *      for information about standard secret key algorithm names.
+     *      Specify "Generic" if the output will be used as the input keying
+     *      material of a key derivation function (KDF).
+     *
+     * @return the shared secret key. The length of the key material
+     *      may be adjusted to be compatible with the specified algorithm,
+     *      regardless of whether the key is extractable. If {@code algorithm}
+     *      is specified as "Generic" and it is supported by the implementation,
+     *      the full shared secret will be returned.
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/KeyAgreement.java:675]: s/will be/is/

File: src/java.base/share/classes/javax/crypto/KeyAgreementSpi.java [hunk 9]
```diff
@@ -205,18 +205,30 @@ protected abstract int engineGenerateSecret(byte[] sharedSecret,
...
+     *      Java Security Standard Algorithm Names Specification</a>
+     *      for information about standard secret key algorithm names.
+     *      Specify "Generic" if the output will be used as the input keying
+     *      material of a key derivation function (KDF).
+     *
+     * @return the shared secret key. The length of the key material
+     *      may be adjusted to be compatible with the specified algorithm,
+     *      regardless of whether the key is extractable. If {@code algorithm}
+     *      is specified as "Generic" and it is supported by the implementation,
+     *      the full shared secret will be returned.
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/KeyAgreementSpi.java:221]: s/will be/is/

**Review by seanjmullan — COMMENTED** [review]
A few more minor comments.

File: src/java.base/share/classes/javax/crypto/KeyAgreement.java (hunk 3 above)
Comment (wangweij) [reply src/java.base/share/classes/javax/crypto/KeyAgreement.java]: Finally, no. KEM changes will be in https://github.com/openjdk/jdk/pull/22929.

**Review by seanjmullan — APPROVED** [review]
